### ISSUE TITLE
Implementing bitrotate.

### ISSUE THREAD

**Issue by pixel27** [eschnett/SIMD.jl#82]
`bitrotate` was added in 1.5, I think, which maps to a `fshl` LLVM command.  However the code has `fshl` (and `fshr`) as a "unary"  operation which I think is incorrect in LLVM land they take 3 parameters:
```
declare <2 x i32> @llvm.fshl.v2i32(<2 x i32> %a, <2 x i32> %b, <2 x i32> %c)
```

Looking at how the code does things, I put this together, and it works, but it might not be as generic as you would like, and I'm not sure if `c` should have declared as type of `T` or not.

```
@generated function Base.bitrotate(a::Vec{N, T}, c) where {N, T<:SIMD.IntegerTypes}
    ff = SIMD.Intrinsics.llvm_name(:fshl, Vec{N, T})
    return :(
        $(Expr(:meta, :inline));
        ccall($ff, llvmcall, Vec{N, T}, (Vec{N, T}, Vec{N, T}, Vec{N, T}), a, a, Vec{N, T}(c))
    )
end
```

Results: 
```
julia> a=Vec{4, UInt32}((0x80000001, 0x80000002, 0x80000003, 0x80000004))
<4 x UInt32>[0x80000001, 0x80000002, 0x80000003, 0x80000004]

julia> bitrotate(a, 1)
<4 x UInt32>[0x00000003, 0x00000005, 0x00000007, 0x00000009]

julia> bitrotate(a, -1)
<4 x UInt32>[0xc0000000, 0x40000001, 0xc0000001, 0x40000002]
```

**Comment by KristofferC** [comment]
> However the code has fshl (and fshr) as a "unary" operation which I think is incorrect in LLVM land they take 3 parameters:

You are right, that looks weird. Those intrinsic should be fixed.

**Comment by eschnett** [comment]
The LLVM intrinsic is more generic than just rotating bits. The way it's implement is indeed a bit rotation. The more generic instruction ("funnel shift") has two arguments, but does something different than a rotation when called with two separate arguments.

We can add a "funnel shift" intrinsic, and can then keep the `bitrorate` one as special case: `bitrotate(a,c) = fsl(a,a,c)`.

`bitrotate` seems to rotate left. The function name doesn't indicate that.

**Comment by pixel27** [comment]
Okay I think I (finally) understand how the code works.  I can take a stab at a fix later today if someone else doesn't pick it up.

**Comment by pixel27** [comment]
This has been fixed by pull request #83.
